new tlcs900 variant: TMP94C241

felipesanches opened the pull request:
Used on the KN5000 driver.

felipesanches commented:
## Before/after comparison

On current git master, the screen gets almost completely black with just a glitch of half a dozen colored pixels at the top line:
![image](https://github.com/user-attachments/assets/403a9a9b-02ff-4c8f-8df5-856f88f3536a)

With the new CPU variant implemented in this PR, we get a good-looking system message regarding control-panel MCU communications error:
![Screenshot From 2025-01-12 04-01-12](https://github.com/user-attachments/assets/0db7d0ed-0613-4259-acdd-1f771151ee29)

felipesanches commented:
This PR is a followup to previous work done at #12684. I believe that I have already addressed all issues pointed out by @cuavas in his previous code review. In this PR I have included only the implementation of the new CPU variant and the small snippets of code in the KN5000 driver that use features on this variant.

The rest of the code for the kn5000 driver will be submitted as separate PRs later to make it easier to review and merge smaller chunks of code.

@wilbertpol and/or @ajrhacker, as you both are listed as authors of other variants of the TLCS900 family, could you please take a look at this PR?

felipesanches commented:
**Note:** The wording of the `** ERROR in CPU data transmission **` message is a bit misleading. I know it is related to the control panel MCUs due to this description in the service manual:

![image](https://github.com/user-attachments/assets/9cc252ea-7dcc-4670-a083-0cc91f2d2314)

felipesanches commented:
To make that work, I'll need to implement UART support in the TMP94C241 CPU device and then use it to attach the MCUs (with HLE initially). But I'd like to land this PR before attempting that, as the amount of code needing review keeps growing.

wilbertpol commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1935 @@
+void tmp94c241_device::treg01_w(offs_t offset, uint8_t data)
+{
+
+    if (BIT(0, offset))
```
This should be `BIT(offset, 0))` ?

wilbertpol commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1935 @@
+void tmp94c241_device::treg23_w(offs_t offset, uint8_t data)
+{
+
+    if (BIT(0, offset))
```
This should be `BIT(offset, 0)` ?

felipesanches replied:
Yes, indeed. I've just pushed a commit fixing it in both places.

wilbertpol commented:
> @wilbertpol and/or @ajrhacker, as you both are listed as authors of other variants of the TLCS900 family, could you please take a look at this PR?

I don't need to be in the copyright-holders section of the tmp94c241.cpp/h files since it is all code made by you.

felipesanches commented:
> > @wilbertpol and/or @ajrhacker, as you both are listed as authors of other variants of the TLCS900 family, could you please take a look at this PR?
> 
> I don't need to be in the copyright-holders section of the tmp94c241.cpp/h files since it is all code made by you.

OK. I've pushed a commit removing your name from the copyright holders list of those files.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1935 @@
+	FF_DONTCARE
+};
+
+void tmp94c241_device::change_timer_flipflop(uint8_t flipflop, uint8_t operation)
```
That method makes no sense.  Look at function 4 on PORT_D, INVERT clears bit 0 (never sets it), SET clears bit 1-7 if bit0=1, CLEAR sets bits 1-7 if bit0=1.  If it's the expected semantics (which I doubt), the names make no sense whatsoever.

Plus it feels like it's always use with constant parameters, which select which line to activate.  Is there a real reason to group them together rather than briging each line up at the calling point?

felipesanches replied:
I've just pushed a re-implementation of this method that I consider to be much better now.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1935 @@
+
+void tmp94c241_device::tlcs900_check_irqs()
+{
+	int irq_vectors[8] = { -1, -1, -1, -1, -1, -1, -1, -1 };
```
Please put variable declarations closer to their first use, C99 is 26 years old at this point and C++ had them even earlier :-)

The int i is particularly egregious, an index variable should escape its loop if and only if its value is used after the loop.

felipesanches replied:
Fixed this now.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1935 @@
+		{
+			m_int_reg[INTE67] &= ~0x08;
+		}
+		m_level[TLCS900_INT3] = level;
```
pretty sure that's wrong

felipesanches replied:
Fixed this now.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1798 @@
+	m_dma_vector[offset] = data;
+}
+
+void tmp94c241_device::b0cs_w(offs_t offset, uint8_t data)
```
Why don't you use 16-bits handlers and COMBINE_DATA ?

felipesanches replied:
I don't know how that works. Can you help me?

galibert replied:
void tmp94c241_device::b0cs_w(offs_t offset, uint16_t data, uint16_t mem_mask)
{
    COMBINE_DATA(&m_block_cs[0]);
}

felipesanches replied:
Also implemented this suggestion. Thanks.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1769 @@
+	m_tffcr = data | 0xcc;
+}
+
+void tmp94c241_device::treg23_w(offs_t offset, uint8_t data)
```
why isn't that two independant handlers?

felipesanches replied:
yes, you're right! Some timers SFRs pack together data for a couple timers, but that's not this case for these ones. I'll fix this!

![Screenshot From 2025-02-09 18-46-55](https://github.com/user-attachments/assets/97f96190-c446-4a78-8ae4-f5bd63d1bec8)

felipesanches replied:
I think I fixed this one as well.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1769 @@
+	m_trdc = data;
+}
+
+void tmp94c241_device::treg45_w(offs_t offset, uint16_t data)
```
Obviously untested code is bad, mmmkay?  offset is going to be 0 or 1, never 2.

felipesanches replied:
`map(0x000090, 0x000093).w(FUNC(tmp94c241_device::treg45_w));`

I expect offset to be between 0 and 3.

felipesanches replied:
oh, I see! data is `uint16_t`... ouch!

felipesanches replied:
I'll make additional commits later tonight. Thanks for finding this bug!

felipesanches replied:
I think I fixed it now.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1768 @@
+}
+
+
+uint8_t tmp94c241_device::cap45_r(offs_t offset)
```
Similarly to write, you can do:
uint16_t tmp94c241_device::cap45_r(offs_t offset)
{
    return m_t16_cap[offset+0];
}

or even a:

template<int timer> uint16_r cap_r()
{
   return m_t16_cap[timer];
}

and change the address map accordingly.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1768 @@
+	map(0x00008b, 0x00008b).w(FUNC(tmp94c241_device::treg3_w));
+	map(0x000090, 0x000091).w(FUNC(tmp94c241_device::treg4_w));
+	map(0x000092, 0x000093).w(FUNC(tmp94c241_device::treg5_w));
+	map(0x000094, 0x000097).r(FUNC(tmp94c241_device::cap45_r));
```
that would become:
map(0x000094, 0x000095).r(FUNC(tmp94c241_device::cap_r<0>));
map(0x000096, 0x000097).r(FUNC(tmp94c241_device::cap_r<1>));

and the following indexes for cap67_r and friends.

galibert commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1768 @@
+	map(0x000082, 0x000082).rw(FUNC(tmp94c241_device::tffcr_r), FUNC(tmp94c241_device::tffcr_w));
+	map(0x000084, 0x000084).rw(FUNC(tmp94c241_device::t01mod_r), FUNC(tmp94c241_device::t01mod_w));
+	map(0x000085, 0x000085).rw(FUNC(tmp94c241_device::t23mod_r), FUNC(tmp94c241_device::t23mod_w));
+	map(0x000088, 0x000088).w(FUNC(tmp94c241_device::treg0_w));
```
there you could have a 8-bit treg03_w and a 16-bits treg45_w both using offset to know which timer to use and remove duplication.  There probably are others too in the map.

The general idea is, if you have essentially the same action on consecutive timers and consecutive addresses, build a handler at the "natural" single timer size, and use offset to select.  If you have essentially the same action, and it's at regular intervals in the memory map and the number of entries is a power of two, using .select().  If you have essentially the same action, but it's all over the place in the memory map, use a template.

felipesanches replied:
done

felipesanches commented:
I've also just rebased this with current master branch.

felipesanches commented:
I've been making commits here to facilitate code-reviews. But not that I've addressed all feedback, perhaps it would be useful for me to git squash it all into a single commit.

felipesanches commented:
I've been making commits here to facilitate code-reviews. But now that I've addressed all feedback, perhaps it would be useful for me to git squash it all into a single commit.

felipesanches commented:
I've squashed all 30 commits into a single one and rebased it against current MAME master branch. The original commits are still available at https://github.com/felipesanches/mame/tree/tlcs900_tmp94c241_before_squashing (just in case it can be useful for the code-reviews).

felipesanches commented:
Also use templates for the two serial channels (squashed-in code corresponding to https://github.com/felipesanches/mame/commit/eeca47ade44155353956c0729a0a9d26b35d75b5)

cuavas commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1520 @@
+
+	/* And here we actually send the value to the corresponding pin */
+	uint8_t new_port_value = 0;
+	switch( flipflop )
+		{
+		case 0x1:
+		case 0x7:
+			new_port_value = m_port_latch[PORT_C] & 0xfe;
+			if (ff_state) new_port_value |= 0x01;
+			port_w<PORT_C>(new_port_value);
+			break;
+		case 0x3:
+		case 0xb:
+			new_port_value = m_port_latch[PORT_C] & 0xfd;
+			if (ff_state) new_port_value |= 0x02;
+			port_w<PORT_C>(new_port_value);
+			break;
+		case 0x4:
+			new_port_value = m_port_latch[PORT_D] & 0xfe;
+			if (ff_state) new_port_value |= 0x01;
+			port_w<PORT_D>(new_port_value);
+			break;
+		case 0x6:
+			new_port_value = m_port_latch[PORT_D] & 0xef;
+			if (ff_state) new_port_value |= 0x10;
+			port_w<PORT_D>(new_port_value);
+			break;
+		case 0x8:
+			new_port_value = m_port_latch[PORT_E] & 0xfe;
+			if (ff_state) new_port_value |= 0x01;
+			port_w<PORT_E>(new_port_value);
+			break;
+		case 0xa:
+			new_port_value = m_port_latch[PORT_E] & 0xef;
+			if (ff_state) new_port_value |= 0x10;
+			port_w<PORT_E>(new_port_value);
+			break;
+		}
```
Same here – braces should be at the same level as the `switch` keyword.

felipesanches replied:
Thanks!

Is this better now? https://github.com/mamedev/mame/pull/13220/commits/77cb08609ab7bad59a4fbbb8fec2d773a0993395

cuavas commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1520 @@
+
+	if ( BIT(m_t8run, 1) ) /* Timer 1 is running */
+	{
+		switch( (m_t01mod >> 2) & 3 ) /* T1_INPUT_CLOCK */
+		{
+		case 0:  /* TO0TRG */
+			break;
+		case 1:  /* T1 */
+			m_timer_change[1] += ( m_timer_pre >> 3 ) - ( old_pre >> 3 );
+			break;
+		case 2:  /* T16 */
+			m_timer_change[1] += ( m_timer_pre >> 7 ) - ( old_pre >> 7 );
+			break;
+		case 3:  /* T256 */
+			m_timer_change[1] += ( m_timer_pre >> 11 ) - ( old_pre >> 11 );
+			break;
+		}
+
+		for( ; m_timer_change[1] > 0; m_timer_change[1]-- )
+		{
+			m_timer_8[1] += 1; /* UPCOUNTER_1 */
+			if ( m_timer_8[1] == m_treg_8[TREG1] )
+			{
+				m_timer_8[1] = 0;
+				m_int_reg[INTET01] |= 0x80;
+				m_check_irqs = 1;
+
+				if ( (m_tffcr & 3) == 3 ) /* "FF1 Invert Enable" && "Invert by timer 1" */
+					change_timer_flipflop( 1, FF_INVERT );
+
+				/* In 16bit timer mode also reset timer 0 */
+				if ( ((m_t01mod >> 6) & 3) == 1 ) /* TO1_OPERATING_MODE == MODE_16BIT_TIMER */
+					m_timer_8[1] = 0;
+			}
+		}
```
This seems to be repeated quite a bit.  Can you encapsulate at least some of it with a local lambda or something?

felipesanches replied:
OK. The remainder of the implementation of timers was generalized (at https://github.com/mamedev/mame/pull/13220/commits/9fcd79c530983d7893120408fa2e5c27257e53bf) with a couple local lambdas: One for the 8bit timers that can also optionally operate in pairs as 16 bit timers. And the other lambda is for the 16-bit only timers.

cuavas commented on src/devices/cpu/tlcs900/tmp94c241.cpp:
```diff
@@ -0,0 +1,1520 @@
+		break;
+
+	case TLCS900_INT4:
+		if (level != m_level[TLCS900_INT4]) {
+			m_level[TLCS900_INT4] = level;
+			if (level == ASSERT_LINE)
+				m_int_reg[INTE45] |= 0x08;
+			else
+				m_int_reg[INTE45] &= ~0x08;
+		}
+		break;
```
Similarly, there are eight almost identical copies of this.  Can you make a local lamba that takes the index to `m_level`, index to `m_int_reg` and bit mask for `m_int_reg` as parameters?

felipesanches replied:
Sure! Do you like it this way? https://github.com/mamedev/mame/pull/13220/commits/38610081dc0d230795c465194f4fda069ef0a7fb

cuavas commented on src/devices/cpu/tlcs900/tmp95c061.cpp:
```diff
@@ -229,6 +229,12 @@ void tmp95c061_device::device_start()
 	save_item(NAME(m_mem_start_mask));
 	save_item(NAME(m_dram_refresh));
 	save_item(NAME(m_dram_access));
+
+	m_nmi_state = CLEAR_LINE;
+	for( int i = 0; i < TLCS900_NUM_INPUTS; i++ )
+	{
+		m_level[i] = CLEAR_LINE;
+	}
```
You really want to do this in `device_resolve_objects` to avoid annoyances when devices push output state on start.

felipesanches commented:
@cuavas, I believe that with the few last commits pushed here, I've addressed all of the things you brought up and this PR is again ready for a review.

felipesanches commented:
Awesome! :-D
